File: 0200_number_of_islands/number_of_islands.js
/**
 * Time:    O(m * n)
 * Space:   O(m * n)
 * @param   {character[][]} grid
 * @return  {number}
 */
const numIslands = grid => {
  let islands = 0
  for (let i = 0; i < grid.length; i++)
    for (let j = 0; j < grid[0].length; j++)
      if (grid[i][j] === "1") islands += dfs(grid, i, j)
  return islands
}

/**
 * @param   {character[][]} grid
 * @param   {number}        row
 * @param   {number}        col
 * @returns {number}
 */
const dfs = (grid, row, col) => {
  if (
    row < 0 ||
    col < 0 ||
    row >= grid.length ||
    col >= grid[0].length ||
    grid[row][col] === "0"
  )
    return (grid[row][col] = "0")
  dfs(grid, row - 1, col)
  dfs(grid, row, col - 1)
  dfs(grid, row + 1, col)
  dfs(grid, row, col + 1)
  return 1
}

File: 0200_number_of_islands/number_of_islands.go
package main

/**
 * Time: O(m * n)
 * Space: O(m * n)
 */
func numIslands(grid [][]byte) int {
	count := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == '1' {
				count += dfs(grid, i, j)
			}
		}
	}
	return count
}

func dfs(grid [][]byte, r int, c int) int {
	if r < 0 || c < 0 || r >= len(grid) || c >= len(grid[0]) || grid[r][c] == '0' {
		return 0
	}
	grid[r][c] = '0'
	dfs(grid, r+1, c)
	dfs(grid, r, c+1)
	dfs(grid, r-1, c)
	dfs(grid, r, c-1)
	return 1
}

func main() {

}
